### PR TITLE
index modifier tool initial impl

### PR DESCRIPTION
**Pull request by miiu96**
- Added a tool that will modify a specified index.

### CODE REVIEW

**Review by dragos-rebegea — DISMISSED** [review]

File: tools/index-modifier/pkg/alterindex/reindex.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+}
+
+func backOff(i int) time.Duration {
+	// A simple exponential delay
+	d := time.Duration(math.Exp2(float64(i))) * time.Second
+	log.Info("elastic: retry backoff", "attempt", i, "sleep duration", d)
+	return d
+}
+
+func CreateIndexModifier(scrollClientAddress, bulkClientAddress string) (*indexModifier, error) {
```
Comment (bogdan-rosianu) [inline tools/index-modifier/pkg/alterindex/reindex.go:33]: add comment

File: tools/index-modifier/pkg/alterindex/reindex.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
+package alterindex
+
+import (
+	"bytes"
+	"fmt"
+	"math"
+	"time"
+
+	client2 "github.com/ElrondNetwork/elastic-indexer-go/client"
```
Comment (bogdan-rosianu) [inline tools/index-modifier/pkg/alterindex/reindex.go:9]: rename `client2` to `indexerClient`

File: tools/index-modifier/pkg/alterindex/reindex.go [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+		return nil, err
+	}
+
+	return &indexModifier{
+		scrollClient: scrollClient,
+		bulkClient:   bulkClient,
+	}, nil
+}
+
+func (im *indexModifier) AlterIndex(indexRead, indexWrite string, modifier func(responseBytes []byte) ([]*bytes.Buffer, error)) error {
```
Comment (bogdan-rosianu) [inline tools/index-modifier/pkg/alterindex/reindex.go:57]: comment

File: tools/index-modifier/pkg/client/client.go [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+	if resp.IsError() && resp.StatusCode != http.StatusNotFound {
+		return fmt.Errorf("error response: %s", resp)
+	}
+
+	return nil
+}
+
+func getBytesFromResponse(res *esapi.Response) ([]byte, error) {
+	if res.IsError() {
+		return nil, fmt.Errorf("error response: %s", res)
```
Comment (bogdan-rosianu) [inline tools/index-modifier/pkg/client/client.go:135]: shouldn't the body be closed even if there is an error?

File: tools/index-modifier/pkg/modifiers/scrsModifier.go [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+		} `json:"hits"`
+	} `json:"hits"`
+}
+
+type scrsModifier struct {
+	pubKeyConverter     core.PubkeyConverter
+	operationDataParser transactions.DataFieldParser
+}
+
+func NewSCRsModifier() (*scrsModifier, error) {
```
Comment (bogdan-rosianu) [inline tools/index-modifier/pkg/modifiers/scrsModifier.go:27]: comment

File: tools/index-modifier/pkg/modifiers/scrsModifier.go [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+		return nil, err
+	}
+
+	return &scrsModifier{
+		pubKeyConverter:     pubKeyConverter,
+		operationDataParser: parser,
+	}, nil
+}
+
+func (sm *scrsModifier) Modify(responseBody []byte) ([]*bytes.Buffer, error) {
```
Comment (bogdan-rosianu) [inline tools/index-modifier/pkg/modifiers/scrsModifier.go:39]: comment

File: tools/index-modifier/pkg/modifiers/txsModifier.go [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+		} `json:"hits"`
+	} `json:"hits"`
+}
+
+type txsModifier struct {
+	pubKeyConverter     core.PubkeyConverter
+	operationDataParser transactions.DataFieldParser
+}
+
+func NewTxsModifier() (*txsModifier, error) {
```
Comment (bogdan-rosianu) [inline tools/index-modifier/pkg/modifiers/txsModifier.go:34]: comment

File: tools/index-modifier/pkg/modifiers/txsModifier.go [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+
+	parser, err := createOperationParser(pubKeyConverter)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return pubKeyConverter, parser, nil
+}
+
+func (tm *txsModifier) Modify(responseBody []byte) ([]*bytes.Buffer, error) {
```
Comment (bogdan-rosianu) [inline tools/index-modifier/pkg/modifiers/txsModifier.go:76]: comment

**Review by bogdan-rosianu — CHANGES_REQUESTED** [review]

File: tools/index-modifier/go.mod [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+require (
+	github.com/ElrondNetwork/elastic-indexer-go v1.1.29-0.20211214150258-c4797dc10a2c
+	github.com/ElrondNetwork/elrond-go-core v1.1.5
+	github.com/ElrondNetwork/elrond-go-logger v1.0.5
+	github.com/elastic/go-elasticsearch/v7 v7.16.0
+	github.com/tidwall/gjson v1.12.1
+)
+
+require (
+	github.com/ElrondNetwork/elrond-vm-common v1.2.7 // indirect
```
[the remark this answers is not in the document]
Comment (miiu96) [reply tools/index-modifier/go.mod:14]: done.

File: tools/index-modifier/pkg/alterindex/interface.go [hunk 10]
```diff
@@ -0,0 +1,15 @@
+package alterindex
+
+import "bytes"
+
+type ScrollClient interface {
```
[the remark this answers is not in the document]
Comment (miiu96) [reply tools/index-modifier/pkg/alterindex/interface.go:5]: added.

File: tools/index-modifier/pkg/alterindex/reindex.go (hunk 1 above)
Comment (miiu96) [reply tools/index-modifier/pkg/alterindex/reindex.go:33]: added

File: tools/index-modifier/pkg/alterindex/reindex.go (hunk 2 above)
Comment (miiu96) [reply tools/index-modifier/pkg/alterindex/reindex.go:9]: done.

File: tools/index-modifier/pkg/alterindex/reindex.go (hunk 3 above)
Comment (miiu96) [reply tools/index-modifier/pkg/alterindex/reindex.go:57]: added

File: tools/index-modifier/pkg/modifiers/scrsModifier.go (hunk 5 above)
Comment (miiu96) [reply tools/index-modifier/pkg/modifiers/scrsModifier.go:27]: done.

File: tools/index-modifier/pkg/modifiers/scrsModifier.go (hunk 6 above)
Comment (miiu96) [reply tools/index-modifier/pkg/modifiers/scrsModifier.go:39]: added

File: tools/index-modifier/pkg/modifiers/txsModifier.go (hunk 7 above)
Comment (miiu96) [reply tools/index-modifier/pkg/modifiers/txsModifier.go:34]: added

File: tools/index-modifier/pkg/modifiers/txsModifier.go (hunk 8 above)
Comment (miiu96) [reply tools/index-modifier/pkg/modifiers/txsModifier.go:76]: added

**Review by bogdan-rosianu — APPROVED** [review]

**Review by dragos-rebegea — APPROVED** [review]

**Pull request merged**